### PR TITLE
[#4] User Service-Basic C&D

### PR DESCRIPTION
**Pull request by sueelly**
## #️⃣ Related Issue

Linked Feature backlog: #4 

## 📝 Purpose of PR

User을 생성 및 삭제하는 로직을 구현한다.

### UserSerivce
- user 생성과 삭제만 진행한다.
- 생성의 경우 OAuth가 정확히 어떻게 돌아가는지 공부하지 못한 상황이라 Email User 기준으로 진행하였다.
- Email User는 생성 시 이메일 인증이 되어있어야만 하고, 삭제 시 비밀번호 인증이 수행된다.
- OAuth User는 삭제 시 이메일 인증을 통해 본인확인이 진행된다.

**SignUp Sequential Diagram**
SignUp (1)

[ 고민되는 사항 ]
- 예외처리는 어디까지 해야하는걸까? email 인증이 완료되었다는 정보만 서버에서 갖고 있을 때, 회원가입 요청을 보내는 이가 인증을 완료한 이와 동일인물인지를 확인하는 로직이 필요할까? 그리고 이에 대한 예외 처리가 필요할까?

### domain module과 application module의 분리 (application module != api module)
- domain module: 순수 자바로 이루어진 도메인 모델을 두어 아무 의존성을 갖지 않는다
- application module: srping-boot-starter 등의 의존성을 갖고 실제 input port와 output port와의 상호작용을 정의하는 Use case와 Output port interface를 정의하고, Use case를 구현한 service를 갖는다.

[ 의도 ]
- 의존성이 필요한 것과 그렇지 않은 곳을 명확히 나눌 수 있다.
- 도메인 모델에서 처리할 수 있는 로직이 많아지게 되는 경우 도메인 로직과 비즈니스 로직을 명확히 구분할 수 있지 않을까 싶었다.

[ 고민이 되는 사항 ]
- core 모듈의 의미: 기존에는 domain module에서 service도 구현하는 방식을 생각했기 때문에 순수 자바로 이루어지고 모두가 공통으로 갖게 되는 것을 core에 정의했다. 하지만 모든 모듈이 domain 모듈에 대한 의존성을 갖고, domain이 순수 자바로 이루어져있다면 분리하는 의미가 있을까? 통합을  고려중이지만 우선 후순위로...
- domain과 application을 나눠야하는가: domain model에서 외부 의존성을 갖지 않도록 (도메인 순수성) 구현을 하면 결국 모델 내에서 처리가능한 로직이 많지 않아졌다. 유효성 검증 수준에서 거의 끝나게 되는데, 이렇게 되는 경우 domain driven이라고 할만한지도 잘 모르겠을 뿐더러 domain과 application을 나누는 의미도 퇴색되는 듯 하다.
- custom exception의 위치: domain model이 빈약해지는만큼 custom exception이 도메인 모듈에서 안 쓰이고 있다. 그리고 대부분의 예외는 service에서 발생한다. 그래서 위치가 고민된다. domain에서 커스텀하게 처리할 무언가가 생길 수 있으니 우선은 domain에 두는 것은 유지한다. 다만 domain 모듈이 아닌 곳에서 새로운 custom exception이 필요할 때마다 domain 모듈을 수정하게 되는 게 옳은 방식인지는 잘 모르겠다. 이것은 domain module이 아니라 common (core) module로 뺀다고 해도 마찬가지...

### UseCase, OutputPort
- 헥사고날 아키텍처를 따른다. application module은 요청을 받아오는 곳이 어딘지 모르며 (RestApi, Websocket, Message etc), 외부의 infra로 무엇을 사용하는지도 전혀 모른다.
- 받아오는 요청을 UseCase로 정의하고, 외부 infra에 보낼 요청을 OutputPort Interface로 정의한다.
- UseCase를 구현하는 것이 Service가 되며, Service는 의존성을 Service와 OutputPort로 갖게 된다. UseCase의 형태로 갖게 되는 경우 UseCase에 없고 service에만 있는 메서드가 필요한 경우 Service를 호출해야하기 때문에 일관성을 맞추기 위해 Service로 통일한다.

### 보완해야하는 부분
- commit convention: commit에서 행위 다음에 ()안에 scope을 넣게 되는데 이를 넣음으로써 섞여있는 커밋들을 나중에 따로 분리할 때, 그리고 한눈에 무엇에 대한 것인지 파악하기가 용이했다. 하지만 해당 scope을 정의하는 게 약간 뒤죽박죽이다. 나는 현재 내가 하고 있는 것의 한단계 위 정도를 나타내는 개념으로 쓰는 것 같다.
- SuperBuilder: domain model은 모두 builder에서 사용하는 생성자를 정의하고, 해당 생성자를 protected/private으로 정의 및 그 안에서 유효성 검증 등의 로직을 추가하였었다. 하지만 SuperBuilder의 경우 정의한 생성자를 사용하지 않는 문제가 발생했다. custom하게 build()를 override하는 방식을 적용하려고 했으나 해결할 수 없는 에러가 계속 발생하여 부득이하게 부모 클래스의 필드를 직접 선언하고, 일반 Builder을 사용하는 형식으로 구현하였다... 

## Checklist

-   [ ] Does commit messages follow the convention?
-   [x] Are variable names brief but descriptive?
-   [x] Is the code clean and easy to understand?
-   [x] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
-   [ ] Does the test successfully test edge and corner cases?

## 💬 Review Requirements (Optional)

Is there anything you can praise about this PR? Start with praise.
Make a review, leaving kind comments and suggesting changes where needed (to resolve the above).

### CODE REVIEW

File: snapnow-application/src/test/java/kr/flab/snapnow/application/user/service/UserSerivceT.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import kr.flab.snapnow.domain.auth.Token;
+import kr.flab.snapnow.domain.user.model.User;
+import kr.flab.snapnow.domain.user.model.userAccount.credential.*;
+import kr.flab.snapnow.application.auth.service.*;
+import kr.flab.snapnow.application.user.fixture.UserFixture;
+import kr.flab.snapnow.application.user.output.UserOutputPort;
+import kr.flab.snapnow.application.email.VerificationType;
+import kr.flab.snapnow.application.email.service.EmailService;
+
+public class UserSerivceT {
```
Comment (f-lab-jason) [inline snapnow-application/src/test/java/kr/flab/snapnow/application/user/service/UserSerivceT.java:24]: T는 뭔가요?

File: snapnow-application/src/test/java/kr/flab/snapnow/application/user/service/UserSerivceT.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    private EmailService emailService;
+    @Mock
+    private UserOutputPort userOutputPort;
+    @InjectMocks
+    private UserService userService;
+
+    @BeforeEach
+    public void setUp() {
+        MockitoAnnotations.openMocks(this);
+    }
```
Comment (f-lab-jason) [inline snapnow-application/src/test/java/kr/flab/snapnow/application/user/service/UserSerivceT.java:40]: 클래스 레벨에 `@ExtendWith`를 사용하시면 이 부분은 딱히 필요없을 것 같아요

**Review by f-lab-jason — COMMENTED** [review]
보통 제반 서비스들(emailService라던지)를 먼저 구현하는데 consumer들을 먼저 구현한 이유가 있을까요? production 코드에 mock up코드를 넣는 일은 거의 없거든요. verification 관련된 코드들도 없어보이구요.

File: snapnow-application/src/test/java/kr/flab/snapnow/application/user/service/UserSerivceT.java (hunk 1 above)
Comment (sueelly) [reply snapnow-application/src/test/java/kr/flab/snapnow/application/user/service/UserSerivceT.java]: Test 약자입니다.
Test를 Test라고 쓰는 곳도 있고 그냥 T라고 쓰는 케이스도 봐서 테스트 작성 시 둘 다 사용되는 줄 알았습니다.
대게 T 하나로는 잘 안쓰나요?

File: snapnow-application/src/test/java/kr/flab/snapnow/application/user/service/UserSerivceT.java (hunk 2 above)
Comment (sueelly) [reply snapnow-application/src/test/java/kr/flab/snapnow/application/user/service/UserSerivceT.java]: 아직 테스트에 익숙치 않아 어떤 형태가 적합한지 잘 모르는 것 같습니다. 수정하도록 하겠습니다!

**Comment by sueelly** [comment]
음... 이게 먼저 구현되어야하는 이유는 딱히 없었습니다.
다만 이제 도메인 모델 -> 서비스 구현의 순서로 진행중이고, 외부 api가 필요하거나 infra가 필요한 것은 후순위로 빼고 있었어서요.
제일 중심이 되는 것을 먼저 넣은 것 같습니다.

보통은 필요한 서비스들을 모두 구현하나보군요... 추후 추가할 서비스들이 잘 작동하는지 테스트를 거친다면, 순서를 거꾸로 해도 문제가 없지 않을까요? 아닌가요?
그러면 테스트에서도 user service단위로만 국한되는 게 아니라 사용되는 것들을 통합하여 테스트를 진행하는 게 옳을까요? consumer의 경우 단위와 통합 둘 다 필요한가요?
